torchmultimodal/models/gpt.py

```diff
@@ -0,0 +1,405 @@
+        return_attn_weights (bool, optional): If ``True``, returns attention probabilities of each transformer
+            layer. Defaults to ``False``.
+        return_hidden_states (bool): If ``True``, returns the embeddings of each transformer layer. Defaults to
+            ``False``.
+    """
+
+    def __init__(
+        self,
+        in_token_emb: nn.Module,
+        out_token_emb: nn.Module,
+        in_pos_emb: nn.Module,
+        out_pos_emb: nn.Module,
+        decoder: nn.Module,
+    ) -> None:
+        super().__init__()
+
+        self.in_token_emb = in_token_emb
+        self.out_token_emb = out_token_emb
+        self.in_pos_emb = in_pos_emb
+        self.out_pos_emb = out_pos_emb
+        self.decoder = decoder
+
+    def forward(
+        self,
+        in_modality: Optional[Tensor] = None,
+        out_modality: Optional[Tensor] = None,
+        in_pos_ids: Optional[Tensor] = None,
+        out_pos_ids: Optional[Tensor] = None,
+        attn_mask: Optional[Tensor] = None,
+        head_mask: Optional[Tensor] = None,
+        use_cache: bool = False,
+        causal: bool = False,
+        return_attn_weights: bool = False,
+        return_hidden_states: bool = False,
+    ) -> TransformerDecoderOutput:
+        if (in_modality is None) and (out_modality is None):
+            raise ValueError(
+                "in_modality and out_modality sequences cannot be both empty"
+            )
+
+        # Since generation is based on the previous data point (autoregressive) where
+        # only one modality is needed at any point along the sequence, either input
+        # or output modality can be None.
+        # Whereas training is done by paralleling all data points so both modalities
+        # should be present. Position ids are optional as they can be derived from
+        # the sequence length of each modality.
+        if in_modality is None:
+            out_pos_ids = self._norm_pos_ids(out_modality, out_pos_ids)
+            x = self._encode(out_modality, out_pos_ids, "out")
+        elif out_modality is None:
+            in_pos_ids = self._norm_pos_ids(in_modality, in_pos_ids)
+            x = self._encode(in_modality, in_pos_ids, "in")
+        else:
+            in_pos_ids = self._norm_pos_ids(in_modality, in_pos_ids)
+            out_pos_ids = self._norm_pos_ids(out_modality, out_pos_ids)
+            x_in = self._encode(in_modality, in_pos_ids, "in")
+            x_out = self._encode(out_modality, out_pos_ids, "out")
+            x = torch.cat((x_in, x_out), dim=1)
+
+        return self.decoder(
+            x,
+            attn_mask,
+            head_mask,
+            use_cache,
+            causal,
+            return_attn_weights,
+            return_hidden_states,
+        )
+
+    def _encode(self, x: Tensor, pos_ids: Tensor, modality: str) -> Tensor:
+        token_emb = getattr(self, f"{modality}_token_emb")
+        pos_emb = getattr(self, f"{modality}_pos_emb")
```

Comment:
Why do we do things this way if we are just gonna use in/out_token/pos_emb in init anyways?

Reply:
Looks good after the latest updates 🙂